File: _streamer/js/stream-manager.js
/**
 * Stream Manager - Handles single stream publishing
 * Manages WHIP streaming to MediaMTX server
 */

import { WHIPClient } from './whip-client.js';
import { updateStatus } from './ui-manager.js';

let whipClient = null;
let isStreaming = false;
let currentStream = null;

export async function handleStartStreaming(mediaStream) {
    if (isStreaming) {
        updateStatus('Already streaming', true);
        return;
    }

    if (!mediaStream || mediaStream.getTracks().length === 0) {
        updateStatus('No media stream available', true);
        return;
    }

    try {
        const whipEndpoint = document.getElementById('whipEndpoint').value.trim();
        if (!whipEndpoint) {
            updateStatus('Please enter a WHIP endpoint URL', true);
            return;
        }

        // Extract base info from WHIP endpoint
        // Expected format: https://server.com/room/streamId/whip
        const whipMatch = whipEndpoint.match(/^(.+)\/([^\/]+)\/([^\/]+)\/whip$/);
        if (!whipMatch) {
            updateStatus('Invalid WHIP endpoint format', true);
            return;
        }

        const [, serverBase, room, streamId] = whipMatch;
        
        // Use the WHIP endpoint directly
        const endpoint = whipEndpoint;

        console.log('Starting stream setup:');
        console.log('Endpoint:', endpoint);

        updateStatus('Starting stream...');

        // Create stream with current settings
        const stream = new MediaStream();
        
        // Add video track
        const videoTrack = mediaStream.getVideoTracks()[0];
        if (videoTrack) {
            stream.addTrack(videoTrack);
            const settings = videoTrack.getSettings();
            console.log('Added video track:', videoTrack.label);
            console.log('📹 Stream Manager: Video track settings:', {
                width: settings.width,
                height: settings.height,
                frameRate: settings.frameRate,
                aspectRatio: settings.aspectRatio,
                deviceId: settings.deviceId
            });

            // Log current UI settings for comparison
            const resolution = document.getElementById('resolution')?.value;
            const bitrate = document.getElementById('videoBitrate')?.value;
            const fps = document.getElementById('fps')?.value;
            console.log('📹 Stream Manager: UI settings:', {
                resolution,
                bitrate: bitrate + ' kbps',
                fps: fps + ' fps'
            });
        }
        
        // Add audio track
        const audioTrack = mediaStream.getAudioTracks()[0];
        if (audioTrack) {
            stream.addTrack(audioTrack);
            console.log('Added audio track:', audioTrack.label);
        }

        whipClient = new WHIPClient(endpoint, (msg) => console.log('Stream:', msg));
        await whipClient.startStream(stream);

        isStreaming = true;
        currentStream = mediaStream;

        // Update UI
        document.getElementById('startButton').disabled = true;
        document.getElementById('stopButton').disabled = false;
        
        // NOTE: We intentionally DO NOT disable video/audio source controls
        // to allow hot-swapping devices during streaming
        console.log('📹 Video and audio device switching remains ENABLED during streaming');
        
        updateStatus('Stream Connected');

        // Store client globally for track replacement
        window.whipClient = whipClient;

        console.log('Streaming started successfully');

        // SEND GREEN ROOM NOTIFICATION THAT STREAM IS NOW LIVE
        if (window.wsConnection && window.participantId) {
            console.log('Notifying green rooms that stream is now live');
            window.wsConnection.emit('stream-live-notification', {
                streamId: window.participantId,
                room: window.room,
                status: 'live-and-ready'
            });
        }

    } catch (error) {
        console.error('Streaming failed:', error);
        updateStatus(`Streaming failed: ${error.message}`, true);
        
        // Cleanup on failure
        cleanup();
    }
}

export function stopStreaming() {
    if (!isStreaming) {
        return;
    }

    updateStatus('Stopping stream...');
    cleanup();
    updateStatus('Stream stopped');
    console.log('Streaming stopped');
}

function cleanup() {
    // Stop WHIP client
    if (whipClient) {
        whipClient.stopStream();
        whipClient = null;
    }

    isStreaming = false;
    currentStream = null;

    // Update UI
    document.getElementById('startButton').disabled = false;
    document.getElementById('stopButton').disabled = true;
    
    // Clear global references
    window.whipClient = null;
}

export function isCurrentlyStreaming() {
    return isStreaming;
}


// Handle track replacement during streaming
export async function replaceTrack(track) {
    if (!isStreaming) return;
    
    try {
        console.log(`Replacing ${track ? track.kind : 'null'} track in stream`);
        
        if (whipClient) {
            await whipClient.replaceTrack(track);
            console.log('Stream track replaced');
        }
        
        console.log('Track replacement completed');
        
    } catch (error) {
        console.error('Error replacing track in stream:', error);
        updateStatus(`Error updating ${track ? track.kind : 'track'}: ${error.message}`, true);
    }
}

File: voodoovideo/StreamManager.swift
import Foundation
import WebRTC
import AVFoundation

/**
 * StreamManager - Manages WHIP streaming lifecycle
 * Mirrors stream-manager.js functionality for native Swift
 */
class StreamManager: ObservableObject {
    
    // MARK: - Published Properties
    @Published var isStreaming = false
    @Published var streamStatus = "Ready"
    @Published var whipEndpoint = ""
    
    // MARK: - Private Properties
    private var whipClient: WHIPClient?
    private var webRTCManager: WebRTCManager?
    private var currentVideoTrack: RTCVideoTrack?
    private var currentAudioTrack: RTCAudioTrack?
    
    // Stream settings (will be synced with VideoManager)
    var videoBitrate: Int = 3000 // kbps
    var frameRate: Int = 30
    var videoCodec: String = "H264"
    var resolution: String = "1080p"
    
    // MARK: - Initialization
    
    init() {
        webRTCManager = WebRTCManager()
        print("🎬 StreamManager: Initialized")
    }
    
    // MARK: - Public Methods
    
    func startStreaming(endpoint: String? = nil) async throws {
        guard !isStreaming else {
            throw StreamError.alreadyStreaming("Already streaming")
        }
        
        // Use provided endpoint or stored endpoint
        let streamEndpoint = endpoint ?? whipEndpoint
        guard !streamEndpoint.isEmpty else {
            throw StreamError.missingEndpoint("WHIP endpoint is required")
        }
        
        print("🎬 StreamManager: Starting stream to endpoint: \(streamEndpoint)")
        updateStatus("Starting stream...")
        
        isStreaming = true
        
        do {
            // Initialize WHIP client
            whipClient = WHIPClient(endpoint: streamEndpoint) { [weak self] message, isError in
                DispatchQueue.main.async {
                    if isError {
                        print("❌ StreamManager: WHIP Client Error - \(message)")
                        self?.updateStatus("Error: \(message)")
                    } else {
                        print("📡 StreamManager: WHIP Client - \(message)")
                        self?.updateStatus(message)
                    }
                }
            }
            
            // Configure WHIP client settings
            whipClient?.videoBitrate = videoBitrate
            whipClient?.frameRate = frameRate
            whipClient?.videoCodec = videoCodec
            
            // Get current tracks from WebRTC manager
            guard let webRTCManager = webRTCManager else {
                throw StreamError.invalidState("WebRTC manager not available")
            }
            
            currentVideoTrack = webRTCManager.getCurrentVideoTrack()
            currentAudioTrack = webRTCManager.getCurrentAudioTrack()
            
            // Start the WHIP stream
            try await whipClient?.startStream(
                videoTrack: currentVideoTrack,
                audioTrack: currentAudioTrack
            )
            
            updateStatus("Stream Connected")
            print("✅ StreamManager: WHIP streaming started successfully")
            
        } catch {
            print("❌ StreamManager: Failed to start streaming - \(error.localizedDescription)")
            isStreaming = false
            updateStatus("Failed: \(error.localizedDescription)")
            throw error
        }
    }
    
    func stopStreaming() {
        guard isStreaming else {
            print("⚠️ StreamManager: Not currently streaming")
            return
        }
        
        print("🛑 StreamManager: Stopping stream...")
        updateStatus("Stopping stream...")
        
        // Stop WHIP client
        whipClient?.stopStream()
        whipClient = nil
        
        // Clear tracks
        currentVideoTrack = nil
        currentAudioTrack = nil
        
        isStreaming = false
        updateStatus("Stream stopped")
        
        print("✅ StreamManager: Streaming stopped")
    }
    
    func replaceVideoTrack(_ track: RTCVideoTrack?) async throws {
        guard isStreaming, let whipClient = whipClient else {
            print("⚠️ StreamManager: Cannot replace video track - not streaming")
            return
        }
        
        print("🔄 StreamManager: Replacing video track")
        
        do {
            try await whipClient.replaceVideoTrack(track)
            currentVideoTrack = track
            print("✅ StreamManager: Video track replaced successfully")
        } catch {
            print("❌ StreamManager: Failed to replace video track - \(error.localizedDescription)")
            throw error
        }
    }
    
    func replaceAudioTrack(_ track: RTCAudioTrack?) async throws {
        guard isStreaming, let whipClient = whipClient else {
            print("⚠️ StreamManager: Cannot replace audio track - not streaming")
            return
        }
        
        print("🔄 StreamManager: Replacing audio track")
        
        do {
            try await whipClient.replaceAudioTrack(track)
            currentAudioTrack = track
            print("✅ StreamManager: Audio track replaced successfully")
        } catch {
            print("❌ StreamManager: Failed to replace audio track - \(error.localizedDescription)")
            throw error
        }
    }
    
    func updateSettings(
        endpoint: String? = nil,
        bitrate: Int? = nil,
        frameRate: Int? = nil,
        codec: String? = nil,
        resolution: String? = nil
    ) {
        if let endpoint = endpoint {
            whipEndpoint = endpoint
        }
        
        if let bitrate = bitrate {
            videoBitrate = bitrate
            whipClient?.videoBitrate = bitrate
        }
        
        if let frameRate = frameRate {
            self.frameRate = frameRate
            whipClient?.frameRate = frameRate
        }
        
        if let codec = codec {
            videoCodec = codec
            whipClient?.videoCodec = codec
        }
        
        if let resolution = resolution {
            self.resolution = resolution
        }
        
        print("🔧 StreamManager: Settings updated")
    }
    
    func processVideoSampleBuffer(_ sampleBuffer: CMSampleBuffer) {
        guard isStreaming else { return }
        
        // Forward to WebRTC manager to create/update video track
        webRTCManager?.processVideoSampleBuffer(sampleBuffer) { [weak self] track in
            Task { @MainActor in
                if self?.currentVideoTrack != track {
                    do {
                        try await self?.replaceVideoTrack(track)
                    } catch {
                        print("❌ StreamManager: Failed to update video track from sample buffer")
                    }
                }
            }
        }
    }
    
    func processAudioSampleBuffer(_ sampleBuffer: CMSampleBuffer) {
        guard isStreaming else { return }
        
        // Forward to WebRTC manager to create/update audio track
        webRTCManager?.processAudioSampleBuffer(sampleBuffer) { [weak self] track in
            Task { @MainActor in
                if self?.currentAudioTrack != track {
                    do {
                        try await self?.replaceAudioTrack(track)
                    } catch {
                        print("❌ StreamManager: Failed to update audio track from sample buffer")
                    }
                }
            }
        }
    }
    
    // MARK: - Private Methods
    
    private func updateStatus(_ status: String) {
        DispatchQueue.main.async { [weak self] in
            self?.streamStatus = status
        }
    }
    
    // MARK: - Ably Integration Callbacks
    
    func notifyStreamLive() {
        // This mirrors the JavaScript functionality where we notify 
        // other components that the stream is live
        print("📡 StreamManager: Stream is now live - notifying other components")
        
        // In the future, we can integrate with Ably here to send notifications
        // similar to the JavaScript version:
        // ablyManager.sendStreamLiveNotification(streamId: participantId, room: room, status: "live-and-ready")
    }
    
    deinit {
        print("💀 StreamManager: Deinitializing")
        if isStreaming {
            stopStreaming()
        }
    }
}

// MARK: - Error Types

enum StreamError: LocalizedError {
    case alreadyStreaming(String)
    case missingEndpoint(String)
    case invalidState(String)
    case trackCreationFailed(String)
    
    var errorDescription: String? {
        switch self {
        case .alreadyStreaming(let message):
            return "Already streaming: \(message)"
        case .missingEndpoint(let message):
            return "Missing endpoint: \(message)"
        case .invalidState(let message):
            return "Invalid state: \(message)"
        case .trackCreationFailed(let message):
            return "Track creation failed: \(message)"
        }
    }
}